### DISCUSSION TITLE
Idea: sequence-expressions

### DISCUSSION THREAD

**Discussion by SebastianMestre** [SebastianMestre/Jasper#213] [Language Feature]
## The problem

It is common in programming to have the initialization of a data structure involve a somewhat long sequence of steps.

As an example, suppose we need to approximate the value of pi for some computation.

```rust
do_work := fn() {
    pi := 0;
    sign := 1.0;
    for (i := 1.0; i < 100.0; i = i + 2.0) {
        pi = pi + sign / i;
        sign = 0.0 - sign;
    }
    pi = pi * 4.0;
    // ...
};
```

In such cases, as seen above, we often end up with intermediate values stored in variables that pollute the local scope, or mutation of the data structure.

A common 'remedy' for this is the extraction of the initialization code to a separate function.

```rust
compute_pi := fn() {
    pi := 0;
    sign := 1.0;
    for (i := 1.0; i < 100.0; i = i + 2.0) {
        pi = pi + sign / i;
        sign = 0.0 - sign;
    }
    return pi * 4.0;
};

do_work := fn() {
    pi := compute_pi();
    // ...
};
```

While this does indeed tidy up the call site, it's not always a good thing to do. For example, this code may be very specific to the caller (e.g. depend on some undocumented invariant), or it may be crucial to understanding the caller code. When this is the case, splitting it off may cause other people to use the function incorrectly, or cause difficulty when reading the caller.

For these reasons, there is a common idiom, in languages that support it, often called IIFE, for "Immediately Invoked Function Expression" (see Javascript and C++). In this idiom, we put the initialization code in a function, but we don't make accessible from the outside, we invoke the expression directly. 

```rust
do_work := fn() {
    pi := fn() {
        pi := 0;
        sign := 1.0;
        for (i := 1.0; i < 100.0; i = i + 2.0) {
            pi = pi + sign / i;
            sign = 0.0 - sign;
        }
        return pi * 4.0;
    }();
    // ...
};
```

While this is a useful idiom, I feel like it's a crutch for a missing language feature, which I will present in the following section.

## Sequence expressions

A sequence expression is an expression that executes a block of code, and evaluates to whatever that block of code returns. This is useful in many cases, and will be even more so once we implement immutability/constants.

Applied to our example, we see the following:

```rust
do_work := fn() {
    pi := seq {
        pi := 0;
        sign := 1.0;
        for (i := 1.0; i < 100.0; i = i + 2.0) {
            pi = pi + sign / i;
            sign = 0.0 - sign;
        }
        return pi * 4.0;
    };
    // ...
};
```

This may not seem like a big improvement, but sequence expressions let us do other interesting things. Namely, they give us a different way to generate the 'short-function' vs 'normal-function' space.

Remember that, in Jasper, we have two ways to write functions. There is the 'short-function', or 'expression-style', and the 'normal-function', or 'statement-style'.

Expression-style functions return whatever their expression evaluates to, and statement-style expressions let us express more involved control flow, in imperative style. Each one is good for different things, so we like having both.

```rust
// expression style
fn(x) => h(g(x));

// statement style
fn(x) {
    a := g(x);
    b := h(a);
    return b;
};
```

Currently, we implement "return whatever their expression evaluates to" in terms of executing a block of code, and using a statement to return a value. This is not nice: We implement a simple thing in terms of a complicated thing.

Back to sequence statements, notice that we have  the following equivalence:

```rust
// classic statement style
fn(x) {
    a := g(x);
    b := h(a);
    return b;
};

// using sequence expressions
fn(x) => seq {
    a := g(x);
    b := h(a);
    return b;
};
```

That is, if we implemented expression-style functions 'natively', along with sequence expressions, we get all the power of statement-style functions, with simpler individual features. Essentially, it's a different way to factor existing features, that gives rise to a new, useful, feature, completely for free.

What do you think?

## Syntax

I have been using `seq { <statements> }` in this post, but that's just a placeholder. If you have an idea for what would be good, go ahead.

**Comment by SebastianMestre** [comment]
pinging @lushoBarlett

**Comment by lushoBarlett** [comment]
It's a nice idea. Like you said, it allows functions to be only one thing, and leaves the "statements and return" functionality to something else, the new feature.

I like it enough to add it, it doesn't have any downsides. Just a nice feature overall. Allows for simplification of some bits of code regarding functions and the like.
